topics/reference_guide/custom_language_support/language_server_protocol.md

```diff
@@ -0,0 +1,183 @@
+
+For more complex cases, you may request to provide a detailed configuration with a dedicated [Settings](settings_guide.md) implementation.
+
+To implement a minimal but fully functional plugin, perform the following steps:
+- Within the `LspServerSupportProvider.fileOpened()` method, spin up the relevant LSP server descriptor, which can decide if the given file is supported by using the `LspServerDescriptor.isSupportedFile()` check method.
+- Tell how to start the server by implementing `LspServerDescriptor.createCommandLine()`.
+
+## Customization
+
+- To fine-tune or disable the implementation of LSP-based features, the plugins may override the corresponding properties of the `LspServerDescriptor` class.
+  See the properties documentation for more details:
+  - `lspGoToDefinitionSupport`
+  - `lspCompletionSupport`
+  - `lspDiagnosticsSupport`
+  - `lspCodeActionsSupport`
+  - `lspCommandsSupport`
+- To handle custom (undocumented) requests and notifications from the LSP server, override `LspServerDescriptor.createLsp4jClient`.
+- To send custom (undocumented) requests and notifications to the LSP server, override `LspServerDescriptor.lsp4jServerClass` and implement the `LspClientNotification` and/or `LspRequest` classes.
+
+See bundled LSP API source code and its code documentation for more information.
+
+## Testing
+
+Utility classes used for testing are not yet extracted for public use, but we'll provide suitable methods for testing LSP-based third-party plugins as soon as possible.
+
+## Troubleshooting
+
+All the IDE and LSP server communication logs are passed to the IDE log file.
+
+To include them for preview, add the following entry to the <control>Help | Diagnostic Tools | Debug Log Settings…</control> configuration dialog:
+
+```
+#com.intellij.platform.lsp
+```
+
+For more information, see the [](ide_infrastructure.md#logging) section.
+
+## Limitations
+
+- The current LSP API implementation assumes that the IDE <-> LSP server communication channel is `stdio`.
+- The IDE doesn't send [workspace/didChangeWatchedFiles](https://microsoft.github.io/language-server-protocol/specification/#workspace_didChangeWatchedFiles) notifications to the server.
+
+## Conclusion
+
+Integrating the Language Server Protocol (LSP) into a plugin for IntelliJ-based IDEs involves a trade-off between simple and fast language support and a complex custom language support plugin with IDE capabilities.
+
+When considering the LSP-based approach, it is important to assess the following criteria for providing a Language Server to end users:
+
+- OS dependency of the Language Server.
+- Availability of the latest version online.
+- Compatibility with breaking changes between versions.
+- Feasibility of requesting the user to provide the Language Server binary path.
+
+> The integration with the Language Server Protocol is created as an extension to the paid IntelliJ-based IDEs.
```

Comment:
I would move it to the top. If someone targets only free IDEs, it will be a waste of time to read it all and get information that it is useless for them.